Using a direct method call with v3client that doesn't go through gRPC to call the Endpoints() method results in a null pointer, causing a panic.

### Bug report criteria

- [X] This bug report is not security related, security issues should be disclosed privately via security@etcd.io.
- [X] This is not a support request or question, support requests or questions should be raised in the etcd [discussion forums](https://github.com/etcd-io/etcd/discussions).
- [X] You have read the etcd [bug reporting guidelines](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/reporting_bugs.md).
- [X] Existing open issues along with etcd [frequently asked questions](https://etcd.io/docs/latest/faq) have been checked and this is not a duplicate.

### What happened?

Using a direct method call with v3client that doesn't go through gRPC to call the Endpoints() method results in a null pointer, causing a panic. The reason is that a variable named `mu` of type `*sync.RWMutex` is not initialized.

## Background

In an effort to minimize Kubernetes (k8s), the apiserver code has been intrusively modified to directly use a v3client for invoking embedded etcd. However, calling the Endpoints() method of etcd's client using this approach leads to a null pointer dereference, causing a panic. The issue stems from an uninitialized variable named `epMu` of type `*sync.RWMutex`.

### Issue Description

The panic occurs as follows:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xa41eac]

goroutine 1 [running]:
sync.(*RWMutex).RLock(...)
        C:/Program Files/Go1.19/go/src/sync/rwmutex.go:69
go.etcd.io/etcd/client/v3.(*Client).Endpoints(0xc0000fe380)
        /go/pkg/mod/go.etcd.io/etcd/client/v3@v3.5.10/client.go:164 +0x4c
main.main()
        D:/01work/code/go/test/main.go:32 +0x1cc    // Endpoints lists the registered endpoints for the client.
```



The null pointer occurs in `client/v3/client.go`, where `c.mu` is not initialized:

```go
// Endpoints lists the registered endpoints for the client.
func (c *Client) Endpoints() []string {
	// copy the slice; protect original endpoints from being changed
	c.mu.RLock()
	defer c.mu.RUnlock()
	eps := make([]string, len(c.cfg.Endpoints))
	copy(eps, c.cfg.Endpoints)
	return eps
}
```



Looking at the initialization code in `server/etcdserver/api/v3client/v3client.go`:

```go
// New creates a clientv3 client that wraps an in-process EtcdServer. Instead
// of making gRPC calls through sockets, the client makes direct function calls
// to the etcd server through its api/v3rpc function interfaces.
func New(s *etcdserver.EtcdServer) *clientv3.Client {
	c := clientv3.NewCtxClient(context.Background(), clientv3.WithZapLogger(s.Logger()))

	kvc := adapter.KvServerToKvClient(v3rpc.NewQuotaKVServer(s))
	c.KV = clientv3.NewKVFromKVClient(kvc, c)

	lc := adapter.LeaseServerToLeaseClient(v3rpc.NewQuotaLeaseServer(s))
	c.Lease = clientv3.NewLeaseFromLeaseClient(lc, c, time.Second)

	wc := adapter.WatchServerToWatchClient(v3rpc.NewWatchServer(s))
	c.Watcher = &watchWrapper{clientv3.NewWatchFromWatchClient(wc, c)}

	mc := adapter.MaintenanceServerToMaintenanceClient(v3rpc.NewMaintenanceServer(s))
	c.Maintenance = clientv3.NewMaintenanceFromMaintenanceClient(mc, c)

	clc := adapter.ClusterServerToClusterClient(v3rpc.NewClusterServer(s))
	c.Cluster = clientv3.NewClusterFromClusterClient(clc, c)

	a := adapter.AuthServerToAuthClient(v3rpc.NewAuthServer(s))
	c.Auth = clientv3.NewAuthFromAuthClient(a, c)

	return c
}

```



And in `client/v3/client.go`, `NewCtxClient` only initializes `lgMu`, not `mu`:

```go
// NewCtxClient creates a client with a context but no underlying grpc
// connection. This is useful for embedded cases that override the
// service interface implementations and do not need connection management.
func NewCtxClient(ctx context.Context, opts ...Option) *Client {
	cctx, cancel := context.WithCancel(ctx)
	c := &Client{ctx: cctx, cancel: cancel, lgMu: new(sync.RWMutex)}
	for _, opt := range opts {
		opt(c)
	}
	if c.lg == nil {
		c.lg = zap.NewNop()
	}
	return c
}
```



### Problem

The main issue is that when using a direct method call approach with a v3client to access an embedded etcd, the necessary mutex (`mu`) in the etcd client is not being initialized. This oversight leads to a runtime panic when attempting to lock this mutex for reading endpoint information.

### Resolution

To resolve this issue, it's necessary to ensure that all required fields, including `mu`, are properly initialized in the etcd client when it's created. This might involve modifying the `New` function in `v3client.go` or the `NewCtxClient` function in `client.go` to include initialization of the `mu` field.

### What did you expect to happen?

Running normally, no null pointer.

### How can we reproduce it (as minimally and precisely as possible)?

demo

```go
package main

import (
	"go.etcd.io/etcd/server/v3/etcdserver/api/v3client"
	"log"
	"time"

	"go.etcd.io/etcd/server/v3/embed"
)

func main() {
	cfg := embed.NewConfig()
	cfg.Dir = "default.etcd"
	cfg.MaxSnapFiles = 10
	cfg.MaxSnapFiles = 10

	e, err := embed.StartEtcd(cfg)
	if err != nil {
		log.Fatal(err)
	}
	select {
	case <-e.Server.ReadyNotify():
		log.Println("Embedded etcd is ready!")
		cli := v3client.New(e.Server)
		endpoints := cli.Endpoints()
		log.Printf("endpoint is %v", endpoints)

	case <-time.After(60 * time.Second):
		e.Server.Stop()
		e.Close()
		log.Fatal("Embedded etcd took too long to start!")
	}

	<-e.Server.StopNotify()
	log.Println("Embedded etcd is stopped")

}

```



### Anything else we need to know?

Hope to solve version 3.5.4.and master

### Etcd version (please run commands below)

test 3.5.4, 3.5.10 still exist

### Etcd configuration (command line flags or environment variables)

see demo

### Etcd debug information (please run commands below, feel free to obfuscate the IP address or FQDN in the output)

no need

### Relevant log output

```Shell
log


{"level":"warn","ts":"2023-11-23T11:33:41.920945+0800","caller":"embed/config.go:676","msg":"Running http and grpc server on single port. This is not recommended for production."}
{"level":"info","ts":"2023-11-23T11:33:41.921003+0800","caller":"embed/etcd.go:127","msg":"configuring peer listeners","listen-peer-urls":["http://localhost:2380"]}
{"level":"info","ts":"2023-11-23T11:33:41.921678+0800","caller":"embed/etcd.go:135","msg":"configuring client listeners","listen-client-urls":["http://localhost:2379"]}
{"level":"info","ts":"2023-11-23T11:33:41.921826+0800","caller":"embed/etcd.go:309","msg":"starting an etcd server","etcd-version":"3.5.10","git-sha":"Not provided (use ./build instead of go build)","go-version":"go1.19.4","go-os":"linux","go-arch":"amd64","max-cpu-set":16,"max-cpu-available":16,"member-initialized":true,"name":"default","data-dir":"default.etcd","wal-dir":"","wal-dir-dedicated":"","member-dir":"default.etcd/member","force-new-cluster":false,"heartbeat-interval":"100ms","election-timeout":"1s","initial-election-tick-advance":true,"snapshot-count":100000,"max-wals":5,"max-snapshots":10,"snapshot-catchup-entries":5000,"initial-advertise-peer-urls":["http://localhost:2380"],"listen-peer-urls":["http://localhost:2380"],"advertise-client-urls":["http://localhost:2379"],"listen-client-urls":["http://localhost:2379"],"listen-metrics-urls":[],"cors":["*"],"host-whitelist":["*"],"initial-cluster":"","initial-cluster-state":"new","initial-cluster-token":"","quota-backend-bytes":2147483648,"max-request-bytes":1572864,"max-concurrent-streams":4294967295,"pre-vote":true,"initial-corrupt-check":false,"corrupt-check-time-interval":"0s","compact-check-time-enabled":false,"compact-check-time-interval":"1m0s","auto-compaction-mode":"","auto-compaction-retention":"0s","auto-compaction-interval":"0s","discovery-url":"","discovery-proxy":"","downgrade-check-interval":"5s"}
{"level":"info","ts":"2023-11-23T11:33:41.92286+0800","caller":"etcdserver/backend.go:81","msg":"opened backend db","path":"default.etcd/member/snap/db","took":"860.481µs"}
{"level":"info","ts":"2023-11-23T11:33:41.923029+0800","caller":"etcdserver/server.go:530","msg":"No snapshot found. Recovering WAL from scratch!"}
{"level":"info","ts":"2023-11-23T11:33:41.924212+0800","caller":"etcdserver/raft.go:530","msg":"restarting local member","cluster-id":"cdf818194e3a8c32","local-member-id":"8e9e05c52164694d","commit-index":31}
{"level":"info","ts":"2023-11-23T11:33:41.924461+0800","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"8e9e05c52164694d switched to configuration voters=()"}
{"level":"info","ts":"2023-11-23T11:33:41.924495+0800","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"8e9e05c52164694d became follower at term 11"}
{"level":"info","ts":"2023-11-23T11:33:41.924507+0800","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"newRaft 8e9e05c52164694d [peers: [], term: 11, commit: 31, applied: 0, lastindex: 31, lastterm: 11]"}
{"level":"warn","ts":"2023-11-23T11:33:41.925618+0800","caller":"auth/store.go:1241","msg":"simple token is not cryptographically signed"}
{"level":"info","ts":"2023-11-23T11:33:41.926804+0800","caller":"mvcc/kvstore.go:407","msg":"kvstore restored","current-rev":10}
{"level":"info","ts":"2023-11-23T11:33:41.928037+0800","caller":"etcdserver/quota.go:94","msg":"enabled backend quota with default value","quota-name":"v3-applier","quota-size-bytes":2147483648,"quota-size":"2.1 GB"}
{"level":"info","ts":"2023-11-23T11:33:41.929179+0800","caller":"etcdserver/server.go:854","msg":"starting etcd server","local-member-id":"8e9e05c52164694d","local-server-version":"3.5.10","cluster-version":"to_be_decided"}
{"level":"info","ts":"2023-11-23T11:33:41.92923+0800","caller":"fileutil/purge.go:44","msg":"started to purge file","dir":"default.etcd/member/snap","suffix":"snap.db","max":10,"interval":"30s"}
{"level":"info","ts":"2023-11-23T11:33:41.929267+0800","caller":"fileutil/purge.go:44","msg":"started to purge file","dir":"default.etcd/member/snap","suffix":"snap","max":10,"interval":"30s"}
{"level":"info","ts":"2023-11-23T11:33:41.929468+0800","caller":"fileutil/purge.go:44","msg":"started to purge file","dir":"default.etcd/member/wal","suffix":"wal","max":5,"interval":"30s"}
{"level":"info","ts":"2023-11-23T11:33:41.929236+0800","caller":"etcdserver/server.go:754","msg":"starting initial election tick advance","election-ticks":10}
{"level":"info","ts":"2023-11-23T11:33:41.92995+0800","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"8e9e05c52164694d switched to configuration voters=(10276657743932975437)"}
{"level":"info","ts":"2023-11-23T11:33:41.930018+0800","caller":"membership/cluster.go:421","msg":"added member","cluster-id":"cdf818194e3a8c32","local-member-id":"8e9e05c52164694d","added-peer-id":"8e9e05c52164694d","added-peer-peer-urls":["http://localhost:2380"]}
{"level":"info","ts":"2023-11-23T11:33:41.930098+0800","caller":"membership/cluster.go:584","msg":"set initial cluster version","cluster-id":"cdf818194e3a8c32","local-member-id":"8e9e05c52164694d","cluster-version":"3.5"}
{"level":"info","ts":"2023-11-23T11:33:41.93012+0800","caller":"api/capability.go:75","msg":"enabled capabilities for version","cluster-version":"3.5"}
{"level":"info","ts":"2023-11-23T11:33:41.93128+0800","caller":"embed/etcd.go:278","msg":"now serving peer/client/metrics","local-member-id":"8e9e05c52164694d","initial-advertise-peer-urls":["http://localhost:2380"],"listen-peer-urls":["http://localhost:2380"],"advertise-client-urls":["http://localhost:2379"],"listen-client-urls":["http://localhost:2379"],"listen-metrics-urls":[]}
{"level":"info","ts":"2023-11-23T11:33:41.931291+0800","caller":"embed/etcd.go:597","msg":"serving peer traffic","address":"127.0.0.1:2380"}
{"level":"info","ts":"2023-11-23T11:33:41.93132+0800","caller":"embed/etcd.go:569","msg":"cmux::serve","address":"127.0.0.1:2380"}
{"level":"info","ts":"2023-11-23T11:33:43.825288+0800","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"8e9e05c52164694d is starting a new election at term 11"}
{"level":"info","ts":"2023-11-23T11:33:43.825331+0800","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"8e9e05c52164694d became pre-candidate at term 11"}
{"level":"info","ts":"2023-11-23T11:33:43.825356+0800","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"8e9e05c52164694d received MsgPreVoteResp from 8e9e05c52164694d at term 11"}
{"level":"info","ts":"2023-11-23T11:33:43.825371+0800","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"8e9e05c52164694d became candidate at term 12"}
{"level":"info","ts":"2023-11-23T11:33:43.825378+0800","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"8e9e05c52164694d received MsgVoteResp from 8e9e05c52164694d at term 12"}
{"level":"info","ts":"2023-11-23T11:33:43.825388+0800","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"8e9e05c52164694d became leader at term 12"}
{"level":"info","ts":"2023-11-23T11:33:43.825396+0800","logger":"raft","caller":"etcdserver/zap_raft.go:77","msg":"raft.node: 8e9e05c52164694d elected leader 8e9e05c52164694d at term 12"}
{"level":"info","ts":"2023-11-23T11:33:43.826476+0800","caller":"etcdserver/server.go:2062","msg":"published local member to cluster through raft","local-member-id":"8e9e05c52164694d","local-member-attributes":"{Name:default ClientURLs:[http://localhost:2379]}","request-path":"/0/members/8e9e05c52164694d/attributes","cluster-id":"cdf818194e3a8c32","publish-timeout":"7s"}
2023/11/23 11:33:43 Embedded etcd is ready!
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xa41eac]

goroutine 1 [running]:
sync.(*RWMutex).RLock(...)
        C:/Program Files/Go1.19/go/src/sync/rwmutex.go:69
go.etcd.io/etcd/client/v3.(*Client).Endpoints(0xc00066e540)
        go/pkg/mod/go.etcd.io/etcd/client/v3@v3.5.10/client.go:164 +0x4c
main.main()
        D:/01work/code/go/test/main.go:25 +0x18a

```
```

--- a/client/v3/client.go
+++ b/client/v3/client.go
@@ -91,7 +91,7 @@ func New(cfg Config) (*Client, error) {
 // service interface implementations and do not need connection management.
 func NewCtxClient(ctx context.Context, opts ...Option) *Client {
 	cctx, cancel := context.WithCancel(ctx)
-	c := &Client{ctx: cctx, cancel: cancel, lgMu: new(sync.RWMutex)}
+	c := &Client{ctx: cctx, cancel: cancel, lgMu: new(sync.RWMutex), epMu: new(sync.RWMutex)}
 	for _, opt := range opts {
 		opt(c)
 	}

